src/modules/landing_target_estimator/LandingTargetEstimator.cpp

```diff
@@ -50,260 +51,1431 @@
+		sensor_ray = R_att * sensor_ray;
 
-		if (fabsf(sensor_ray(2)) < 1e-6f) {
-			// z component of measurement unsafe, don't use this measurement
-			return;
+		//TODO: rotate measurement uncertainty S_att & R_att
+
+		// z component of measurement safe, use this measurement
+		if (fabsf(sensor_ray(2)) > 1e-6f) {
+
+			float dist_z = _dist_bottom - _param_ltest_sens_pos_z.get();
+
+			// scale the ray s.t. the z component has length of _uncertainty_scale
+			float rel_pos_x = sensor_ray(0) / sensor_ray(2) * dist_z;
+			float rel_pos_y = sensor_ray(1) / sensor_ray(2) * dist_z;
+			float rel_pos_z = dist_z;
+
+			// Adjust relative position according to sensor offset
+			rel_pos_x += _param_ltest_sens_pos_x.get();
+			rel_pos_y += _param_ltest_sens_pos_y.get();
+
+			//TODO: For coupled dynamics: Cholesky decomposition as for vision obs to uncorrelate noise (find matrix T that diagonalizes R) z' = T*z; H' = T*H
+
+			// Fill the observations for the irlock sensor
+			obs->timestamp = irlock_report.timestamp;
+
+			if (_target_model == TargetModel::FullPoseCoupled) {
+				// State: [rx, ry, rz, ... ]
+				obs->meas_h_xyz(0, 0) = 1; // x direction
+				obs->meas_h_xyz(1, 1) = 1; // y direction
+				obs->meas_h_xyz(2, 2) = 1; // z direction
+
+			} else {
+				// State: [r, r_dot, ...] --> x, y, z directions are the same (decoupled)
+				obs->meas_h_xyz(0, 0) = 1;
+				obs->meas_h_xyz(1, 0) = 1;
+				obs->meas_h_xyz(2, 0) = 1;
+			}
+
+			obs->meas_xyz(0) = rel_pos_x;
+			obs->meas_xyz(1) = rel_pos_y;
+			obs->meas_xyz(2) = rel_pos_z;
+
+			obs->updated_xyz.setAll(true);
+
+			float measurement_uncertainty = _meas_unc * dist_z * dist_z;
+
+			obs->meas_unc_xyz(0) = measurement_uncertainty;
+			obs->meas_unc_xyz(1) = measurement_uncertainty;
+			obs->meas_unc_xyz(2) = measurement_uncertainty;
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool LandingTargetEstimator::fuse_meas(const Vector3f vehicle_acc_ned, const targetObsPos target_pos_obs)
+{
+	estimator_aid_source_3d_s target_innov;
+	Vector<bool, 3> meas_xyz_fused{};
+	bool all_directions_fused = false;
+	Vector<float, 12> meas_h_row;
+
+	// Number of direction:  x,y,z for all filters excep for the horizontal filter: x,y
+	int nb_update_directions = (_target_model == TargetModel::Horizontal) ? 2 : 3;
+
+	// Compute the measurement's time delay (difference between state and measurement time on validity)
+	float dt_sync = (_last_predict - target_pos_obs.timestamp);
+
+	if (dt_sync > measurement_valid_TIMEOUT_US) {
+
+		PX4_INFO("Obs i = %d rejected because too old. Time sync: %.2f [seconds] > timeout: %.2f [seconds]",
+			 target_pos_obs.type,
+			 (double)(dt_sync / SEC2USEC), (double)(measurement_valid_TIMEOUT_US / SEC2USEC));
+
+		// No measurement update, set to false
+		for (int k = 0; k < 3; k++) {
+			target_innov.fusion_enabled[k] = false;
+			target_innov.fused[k] = false;
+		}
+
+	} else {
+
+		// Convert time sync to seconds
+		dt_sync = dt_sync / SEC2USEC;
```

Comment:
use a new variable instead of changing it's meaning, e.g.: `const float dt_sync_s = dt_sync / SEC2USEC` and maybe rename `dt_sync` to `dt_sync_us`

Reply:
Done in https://github.com/JonasPerolini/PX4-Autopilot/commit/ea0c0113b1a7fbe7f4465e657dd69ad93a59c675